docs(RFC): disruption controls by reason (#936)
kubernetes-sigs/karpenter: designs/disruption-controls-by-reason.md

# Disruption Controls By Reason 
## User Scenarios 
1. Users need the capability to schedule upgrades only during business hours or within more restricted time windows. Additionally, they require a system that doesn't compromise the cost savings from consolidation when upgrades are blocked due to drift.
2. High-Frequency Trading (HFT) firms require full compute capacity during specific operational hours, making it imperative to avoid scale-down requests for consolidation during these periods. However, outside these hours, scale-downs are acceptable.
See Less Made Up Scenarios here: 
- https://github.com/kubernetes-sigs/karpenter/issues/924 
- https://github.com/kubernetes-sigs/karpenter/issues/753#issuecomment-1790110838
- https://github.com/kubernetes-sigs/karpenter/issues/672

## Known Requirements 
**Reason and Budget Definition:** Users should be able to define an reason and a corresponding budget(s).
**Supported Reasons:** All disruption Reasons affected by the current Budgets implementation (Consolidation, Emptiness, Expiration, Drift) should be supported. 

> **Default Behavior for Unspecified Reasons:** Budgets should continue to support a default behavior for all disruption reasons. If an reason is unspecified, it is assumed to apply to all reasons. If a reason is unspecified, we apply the budget to be shared by all reasons that are unspecified. 

Reviewer: nit: this section is a bit hard to parse.

> If an reason is unspecified, it is assumed to apply to all reasons. If a reason is unspecified, we apply the budget to be shared by all reasons that are unspecified.

becomes

> If a reason is unspecified:
>* it is assumed to apply to all reasons
>* we apply the budget to be shared by all reasons that are unspecified

*Or alternatively*
> If an reason is unspecified, it is assumed to apply to all reasons. If a reason is unspecified, we apply the budget to be shared by all reasons that are unspecified.

becomes
> If an reason is unspecified, it is assumed to apply to all reasons. We will apply the budget to be shared by all reasons that are unspecified.

Author: yeah good catch

## API Design
### Approach A: Add a reason field to disruption Budgets 
This approach outlines a simple api change to the v1beta1 nodepool api to allow disruption budgets to specify a list of disruption methods. 

### Proposed Spec
Add a simple field "reasons" is proposed to be added to the budgets. 
```go
// Budget defines when Karpenter will restrict the
// number of Node Claims that can be terminating simultaneously.
type Budget struct {
      // Reasons is a list of methods for disruption that apply to this budget. If Reasons is not set, this budget applies to all methods.
      // If a reason == "default", it will apply to all reasons that don't have an active budget. If a reason is set, it will only apply to that method. If multiple reasons are specified,
      // this budget will apply to all of them. If a budget is not specified for a method, the default budget will be used.
      // allowed reasons are "default", "consolidation", "expiration", "emptiness", "drift"
      // +kubebuilder:validation:MaxItems=5
      // +kubebuilder:validation:Enum:={"default","consolidation","expiration","emptiness","drift"}

>       // +optional

Author: #### Q: Should default be specifed only as omitted? Or should users be able to define default? 
We talked about having "default" as an action case. Meaning that if I specify a budget for consolidation and Drift, then a separate budget with an empty reason in the same active window will be taken as the default number of disruptable nodes in a given window for all actions that were not defined. Should we also be able to define default as a reason explicitly? 

The answer is yes to reduce redundancy. 

I can define a single budget to cover all reasons for disruption like so 

```
  - nodes: 10
  reasons: [Drift, Consolidation, Default]
  schedule: "* * * * *"
```
Rather than having to specify two budgets like 
```
  - nodes: 10
  reasons: [Drift, Consolidation]
  schedule: "* * * * *"
  - nodes: 10
  schedule: "* * * * *"
```

To communicate the same thing to karpenter. So we should allow users to specify default directly and explicitly, alongside having the behavior for default from specifying like 
```
  - nodes: 10
  schedule: "* * * * *"
```

### Approach B: Defining Per Reason Controls  
Ideally, we could move all generic controls that easily map into other reasons into one set of reason controls, this applies to budgets and other various disruption controls that could be more generic. 
### Proposed Spec 
```go
type Disruption struct {
    Default	  DisruptionSpec `json:defaults"`
    Consolidation ConsolidationSpec `json:"consolidation"`
    Drift         DriftSpec         `json:"drift"`
    Expiration    ExpirationSpec    `json:"expiration"`
    Emptiness     EmptinessSpec     `json:"emptiness"`
}

type DisruptionCommonSpec struct {
    DisruptAfter string   `json:"disruptAfter"`
    Budgets      []Budget `json:"budgets"`
}

type ConsolidationSpec struct {
    DisruptionCommonSpec
    ConsolidationPolicy string `json:"consolidationPolicy"`
}

type DriftSpec struct {
    DisruptionCommonSpec
}

type ExpirationSpec struct {
    DisruptionCommonSpec
}

type EmptinessSpec struct {
    DisruptionCommonSpec
}

type Budget struct {
    Nodes    string  `json:"nodes"`
    Schedule *string `json:"schedule,omitempty"`
    Duration *string `json:"duration,omitempty"`

>     Reasons []string 

Author: We have three options for the name of the field. 

1. Actions 
2. Reasons 
3. Methods 

I removed all the talk from the doc of drift reasons, and we aren't designing it right now, but also this API will be used directly after to represent them or other reasons. 

REasons to me makes more sense than Methods because its more extensible, and methods directly reference the primary 4 methods in the karpenter disruption controller.

>     Reasons []string 

Author: Rather than changing the API in the future, it would be good to see a general list of arbitrary reasons that can be extended. 

Allthough we may see that we don't want to reuse the same stucture for that design. Lets just call it methods and if we need to change it we can.

>     Reasons []string 

Reviewer 2: Fwiw I'm +1 for "reasons". "Methods" doesn't read as well in plain English. It's essentially noun (reason=expired) vs verb (method=expire), and declarative API policies tend towards nouns.

>     Reasons []string 

Author: Lets have a vote at the next WG, I am leaning more on reasons as welll after reflecting a bit on it.

>     Reasons []string 

Reviewer 3: Ah I thought reasons referred to a specific per-method reason of disrupting, like NodePoolDrift or Underutilized for Consolidation. We use methods in code, but I agree reasons semantically make more sense to refer to these. In docs we call these methods. Wherever we land on this, it'd be good to update that language throughout (obviously where it makes sense)

>     Reasons []string 

Author: Ah good call

>     Reasons []string 

Reviewer 4: > Wherever we land on this, it'd be good to update that language throughout

FWIW, I think that users probably don't care about method when there's an overlap between consolidation and emptiness. I really only care, "is my node fully empty or only partially empty". There's perhaps an argument to be made that we should change our metrics and logging to align with "reason" in the new API rather than the other way around.

}

```
#### Example 

```yaml 
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: example-nodepool
spec:
  disruption:
    defaults:
      budgets: 
        - nodes: 10% 
          schedule: "0 0 1 * *"
          duration: 1h 
    consolidation:
      consolidationPolicy: WhenUnderutilized
      disruptAfter: "30m"
    drift:
      budgets:
        - nodes: "20%"
          schedule: "0 0 1 * *"
          duration: "1h"
        - nodes: "10%"
          schedule: "0 0 * * 0"
          duration: "2h"
        - nodes: "50%" 
          schedule: "@monthly"
    expiration:
      disruptAfter: "Never"
```
#### Considerations 

> Some of the API choices for a given reason seem to follow a similar pattern. These include ConsolidateAfter, ExpireAfter, and there are discussions about introducing a global DisruptAfter. Moreover, when discussing disruption budgets, we talk about adding behavior for each reason. It appears there is a need for disruption controls within the budgets for each reason, not just overall.

Author: https://github.com/kubernetes-sigs/karpenter/issues/920#issuecomment-1874946264 is the only public mention I recall. Others were dms and discussions. 

Would be interesting to more formally design and document a disruptAfter: x minutes.

## Clarifying Expected Behavior 
### Q: How should Karpenter handle the default or undefined reason case? 
The current design involves specifying a specific number of disruptable nodes per reason, which can complicate the disruption lifecycle. For example, if there's a 10-node budget for "Drift" and a separate 10-node budget for "Consolidation", but a 15-node budget for "default" determining which nodes will get disrupted becomes unclear. Would it be 10 nodes for "Drift" and 5 nodes for "Consolidation"?

We could consider treating an undefined reason as a budget for all disruption reasons except for those with explicitly defined budgets. In this scenario, if a user specifies a disruption budget like this:

```yaml
spec: # This is not a complete NodePool Spec.
  disruption:
    budgets:
    - schedule: "* * * * *"
      reason: Consolidation
      nodes: 10
    - schedule: "* * * * *"
      reason: Drift
      nodes: 10
    # For all other reasons , only allow 5 nodes to be disrupted at a time
    - nodes: 5
      schedule: "* * * * *"
```

> It means that "Consolidation" and "Drift" reasons have specific budgets of 10 nodes each, while all other reasons (e.g., expiration and emptiness) share a common budget of 5 nodes. This approach simplifies the configuration but has one limitation: it may not allow the execution of other disruption reasons if a specific reason exhausts the budget. This is a problem with the existing design for disruption budgets.

Author: I wrote an option in the doc for using both "All" and "default" lmk your thoughts

> It means that "Consolidation" and "Drift" reasons have specific budgets of 10 nodes each, while all other reasons (e.g., expiration and emptiness) share a common budget of 5 nodes. This approach simplifies the configuration but has one limitation: it may not allow the execution of other disruption reasons if a specific reason exhausts the budget. This is a problem with the existing design for disruption budgets.

Author: https://github.com/kubernetes-sigs/karpenter/blob/17de53492c9ee6d167f8e72cd7b0088cc1d62ce9/designs/disruption-controls-by-reason.md#q-should-default-be-specifed-only-as-omitted-or-should-users-be-able-to-define-default

There are two ways for the users to get around this behavior. 
1. If you need guaranteed disruption for a particular action, you can just specify that action in a budget.  
2. We could allow some mechanism for the users to control the ordering of the disruption actions.

#### Q: Should users be able to change the order that disruption reasons are executed in to solve this problem? 
The answer is no, this makes it harder for cluster operators to understand behavior. It also doesn't elegantly fit into karpenters per nodepool controls. Defining it in the nodepool would mean you have multiple nodepools with different orderings, which is diffcult. Karpenter today does not provide an easy way via the CRDS to define per cluster level controls.  

> #### Q: Should Budget "Reasons" allow for specifying more than one? Via an array? 

Author: Was thinking similar thoughts, can move to its own API section

> #### Q: Should Budget "Reasons" allow for specifying more than one? Via an array? 

Author: Moved it to its own design section 🫡

   - [Q: Default or Undefined Reason Case Handling](#q-default-or-undefined-reason-case-handling)
   - [Q: Calculation of Allowed Disruptions in a Multi-Reason World](#q-calculation-of-allowed-disruptions-in-a-multi-reason-world)
   - [Q: Handling an Unspecified Default Reason](#q-handling-an-unspecified-default-reason)
   - [Q: Definition of Default in Budgets](#q-definition-of-default-in-budgets)
4. [API Design](#api-design)
   - [Approach A: Extending the Budget API to Specify a Reason](#approach-a-extending-the-budget-api-to-specify-a-reason)
      * [List Approach: Multiple Reasons per Budget](#list-approach-multiple-reasons-per-budget)
      * [Single Reason Approach: One Reason per Budget](#single-reason-approach-one-reason-per-budget)
      * [Pros and Cons for Both Approaches](#pros-and-cons-for-both-approaches)
      * [Preferred Option: List Approach](#preferred-option-list-approach)
   - [Approach B: Defining Per Reason Controls](#approach-b-defining-per-reason-controls)
      * [Pros and Cons](#pros-and-cons)
   - [API Design Conclusion: Preferred Design](#api-design-conclusion-preferred-design)

## User Scenarios 

> 1. Users need the capability to schedule upgrades only during business hours or within more restricted time windows. Additionally, they require a system that doesn't compromise the cost savings from consolidation when upgrades are blocked due to drift.

Reviewer 4: The one question that I have on this scenario: If we are consolidated with a replacement, the replacement node is still going to be on a new version and is going to be effectively using the new configuration, so blocking drift isn't really doing anything here in that sense. We are really just blocking drift if and only if the nodes aren't underutilized enough. Is that enough or even what the user wants? I'm really trying to dig in here to understand why a user would block a node upgrade and if we are really fulfilling their use-case by blocking this node upgrade on some operations but not others.

Author: Yeah thats a fair point. Again this flaw pops up where there is some bleed where the user specifies: "don't disrupt due to drifted" but then those nodes will still be liable for other disruption. 

It doesn't succeed very well at blocking NodeImageUpgrade and the more I think about it the more I think that these controls do not meet the upgrade user story in a complete way. 

CC: @charliedmcb  who is cooking up some additional more focused upgrade controls for karpenter.

2. Users want to minimize workload disruptions during business hours but still want to be able to delete empty nodes throughout the day.  That is, empty can run all day, while limiting cost savings and upgrades due to drift to non-business hours only.

See Less Made Up Scenarios here: 

> - https://github.com/kubernetes-sigs/karpenter/issues/924 

Reviewer 4: This is great! More of our design docs should reference out to issues like this; great model for others!

- https://github.com/kubernetes-sigs/karpenter/issues/753#issuecomment-1790110838
- https://github.com/kubernetes-sigs/karpenter/issues/672

## Clarifying the requirements and behavior 
**Reason and Budget Definition:** Users should be able to define an reason and a corresponding budget(s).
**Supported Reasons:** All disruption Reasons affected by the current Budgets implementation (underutilized, empty, expired, drifted) should be supported. 
**Default Behavior for Unspecified Reasons:** Budgets should continue to support a default behavior for all disruption reasons. 

### Q: How should Karpenter handle the default or undefined reason case? 

> If a budget reason is unspecified like budgets[2], we will assume this budget applies to all actions that are not specified 

Reviewer 3: I can't tell if this is an artifact of a previous version or you decided to 1-index this :D 
```suggestion
If a budget reason is unspecified like budgets[1], we will assume this budget applies to all actions that are not specified 
```

Author: Yeah this doc went through too many adhoc adjustments. Need to go through and properly clean up.

> If a budget reason is unspecified like budgets[1], we will assume this budget applies to all actions that are not specified 

Reviewer 4: Is this the correct behavior? Or should we just assume that it applies to all reasons? Meaning, if this budget was more restrictive, it would just apply to everything and the less restrive budget just wouldn't apply. Think if `budgets[1]` was:

```yaml
budgets:
- nodes: 5
```

Reviewer 3: Based off other convos, I think this is just an artifact from a previous revision

Reviewer 5: I was wondering about this as well. How this is current worded when specifying by reason is an override value which will take precedence over a generic budget even if the generic budget is lower. 

That seems somewhat counter to the "use most restrictive budget", but also does seems potentially alright as I could see in better aligning with real world usage? You might want to have a lower budget generally, but possibly some exemptions to that case for certain reasons? Would be useful to have real world customer desired usage.

Author: > I was wondering about this as well. How this is current worded when specifying by reason is an override value which will take precedence over a generic budget even if the generic budget is lower. 

This was my intent. We should use the most restrictive budget per reason. The default budget should only apply to actions not specifed. Is this behavior confusing?  We agreed to not have empty reason mean all actions but instead the default for other actions. 

This budget allows us to say, don't disrupt due to drift or utilization, but feel free to disrupt for emptiness and for expiration as much as you want as I as a user want to be able to have larger disruption for these reasons.

Author: 2. AllowedDisruptionByReason = minNodeCountOfActiveBudget[reason] - unhealthyNodes - totalDisruptingNodes

we still consider all total disruptions, but it would be the min defined by each reason in the budget declaration

Author: would you propose the below equation instead? 

AllowedDisruptionByReason = min(minNodeCount[reason], minNodeCount["default"]) - unhealthyNodes - totalDisruptingNodes

Author: Let me add this as an equation 3 and we write a thorough pros and cons list going through the 3 equations going through scenarios specified here: https://github.com/kubernetes-sigs/karpenter/pull/936/files#r1503748815

Author: After outlining a couple of scenarios, lets go with the inverse of what this says above.  Will push those scenarios 

Take a look at the following budgets. The first budget is easier to reason about and more elegant. Lets ditch this idea. 

# User Story 

Customers who roll out aggressive node image upgrades don't want to roll the nodes due to drift, but want to use node expiration to force users onto compliant node images eventually. The scenario here is an ML Training workload. Ideally we want to let it bake as long as possible, but not at the cost of compromising the security of the cluster.

```yaml 
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: worker-pool 
spec: # This is not a complete NodePool Spec.
  disruption:
    consolidationPolicy: WhenEmpty 
    expireAfter: 7d 
    budgets:
    - nodes: 0
      reasons: [Drifted] 
    - nodes: "100%" // All Disruption Other than drift(Expiration and Empty Nodes) are allowed
```

We could also allow a semantic like this
```yaml
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: worker-pool 
spec: # This is not a complete NodePool Spec.
  disruption:
    consolidationPolicy: WhenEmpty 
    expireAfter: 7d 
    budgets:
    - nodes: "100%" # Always allow Expiration and Emptiness to disrupt 
      reasons: [Expired, Empty]
    - nodes: 0
```
The second semantic may not allow us to define equation 3, but the semantic requires extra work and doesn't follow the original least disruptive principle of the budgets.

Reviewer 5: These are some good detailed comments.

> AllowedDisruptionByReason = min(minNodeCount[reason], minNodeCount["default"]) - unhealthyNodes - totalDisruptingNodes

^ this is a great way of summarizing what I mean.

As the examples you laid out something like the following wouldn't work:
> budgets:
>    \- nodes: "100%" # Always allow Expiration and Emptiness to disrupt 
>      reasons: [Expired, Empty]
>    \- nodes: 0

As by equation 3 all scenarios would be 0. The override scenario pattern, and equation 2 feels the most natural to me, and potentially useful, but wanted to make sure these was at least called out, and discussed as its breaking the "always use the most restrictive" budget if we go with equation 2 over equation 3.

Reviewer 5: A potential way to address this too would be to have a "default" like equation 2, but also support an "All" option? That way if people do want to make sure they have a safe override for example where no disruptions happen they can use the "All" scenario.

### Q: How should allowed disruptions be calculated in a multi-reason world? 
The calculation of allowed disruptions in a system with multiple disruption reasons has become more intricate following the introduction of Disruption Budget Reasons. Previously, the formula was straightforward:

AllowedDisruptions = minNodeCountOfAnActiveBudget - unhealthyNodes - totalDisruptingNodes.

When calculating by reason, two potential equations emerge:
1. AllowedDisruptionByReason = minNodeCountOfActiveBudget[reason] - unhealthyNodes - totalDisruptingNodes[reason]
2. AllowedDisruptionByReason = minNodeCountOfActiveBudget[reason] - unhealthyNodes - totalDisruptingNodes
The second equation is the one we should opt into.

Take this budget as an example
```yaml
budgets: 
  - nodes: 15 
    reasons: [drifted, underutilized]
    schedule: "* * * * *"
  - nodes: 10
    reasons: [drifted]
    schedule: "* * * * *"
  - nodes: 5 
    schedule: "* * * * *" 
```
First, calculate minNodeCountOfActiveBudget. Given there are two budgets for drifted, we select the lower number:

```
minNodeCountOfActiveBudget = {
  drifted: 10 
  underutilized: 15 

>   Default: 5

Reviewer 3: one question on the explanation here:

In this specific example, the nodes: 5 will always be the budget, since it's always enabled, and effectively is the same as `reasons: [*]`. Can we change up the example here with maybe changing this to a percent so it's possible to see that all the different budgets can be active at different points?

Reviewer 4: +1, I think I had a similar question up above where we are making the determination on what not specifying the reason means.

Author: >  , and effectively is the same as reasons: [*] 
"" != "all" its all that are undefined was my intention. Do you expect that if I have 

minNodeCountOfActiveBudget = {
  drifted: 10 
  underutilized: 15 
  Default: 5
 }
 
 That if the 5 Node declaration should bound ALL disruption reasons?  That it should floor all reasons? That equation is slightly different from what I am proposing.

}
```

In the first equation, tracking the number of nodes disrupted by each reason is required. For example:
```
disrupting = {
  drifted: 3 
  underutilized: 6
  Expriation: 3
  empty: 2
}
```

Assuming zero unhealthy nodes:
**First Equation**
- drifted = 10 - 3: Disruption allowed as drifted bucket isn't full.
- underutilized = 15 - 6: Disruption allowed since 15 - 6 > 0.
- Default = 5 - expired - empty = 0: No disruption allowed for reasons other than consolidation and drift.

**Second Equation**
- Total Disrupting = 3 + 6 + 5 = 14.
- drifted = 10 - 14: No disruption allowed due to drift.
- underutilized = 15 - 14: Disruption allowed for at least one node due to consolidation.
- Default: 5 - 14: No disruption allowed for other methods.

#### Decision

> The second equation simplifies reasoning for cluster operators. A node can be tagged for multiple disruption reasons (e.g., empty, drifted, underutilized), making it challenging to precisely calculate in-flight disruptions. The second equation streamlines this process by providing a clearer view of the overall disruption impact.

Reviewer 4: The other thing that I think that we should think about here is the user's intention when they say, "I don't want more than x nodes drifted at a given time." It may be worth formalizing set-ups that we actually expect users to configure in the doc to be able to better reason about what makes sense here. Off the top of my head, I can see users wanting to do the following things:
1. Fully blocking drift/expiration while allowing consolidation 
2. Fully blocking drift/expiration while only allowing emptiness
3. Allowing drift/expiration to occur one-at-a-time while allowing consolidation to be more aggressive
...

Reviewer 4: I still think that the second equation makes the most sense just from an "ability to reason about it" standpoint, but I just want to double check and make sure that all configurations that we expect to happen are covered by this calculation.

Author: @jonathan-innis  let me know what you think about the scenarios section I added and if its a bit too verbose. Basically went through 7 scenarios excercising the budgets by reason with a configured nodepool.

> ### Q: How should we handle an unspecified default reason? 

Reviewer 4: ```suggestion
### Q: How should we handle an unspecified reason when others are specified? 
```

I was a little unclear on this when I first read it. Maybe tightening up the language of the question could help here. From what I understand, you are asking how do we handle the case for expiry below since it's not specified in the budget.

```yaml
budgets: 
  - nodes: 10
    reasons: [drift, underutilized]
    schedule: "* * * * *"
  - nodes: 5 
    reasons: [empty] 
    schedule: "* * * * *" 
```
In the case of a budget like above, default is undefined. Should karpenter assume the user doesn't want to disrupt any other reasons? Or should we assume that if a default is unspecified, they want us to disrupt anyway?  

> The intuitive options if there is no active default budget is to allow disruption of either 0 or total number of nodes(meaning unbounded disruption).

Reviewer 4: The other point of concern here is that once you are specifying budgets with reasons, we can't default you to 10% like we do generally when you don't specify a budget. I think this is fine, considering if you are getting this into the weeds with your budgets, you should know that you need to set a budget that covers the cases that you didn't explicitly state, but it's worth calling out.

In either case, unbounded strikes me as the right behavior.

Reviewer 3: This is the expected behavior today as well. If you don't have a budget with a schedule matching the current time, then it'll be unbounded.

Author: @njtran When i test this on the current karpenter core v0.34.1, seems that not specifying a budget, i get 10% budget applied by default. 

Applying this nodepool with no budget specified https://github.com/Azure/karpenter-provider-azure/blob/main/examples/v1beta1/general-purpose.yaml, we get 10%, and disruption is blocked due to this budget. 

```
~/dev/karp/karpenter-provider-azure (bsoghigian/v0.34-clean*) » k get nodepools.karpenter.sh general-purpose -o yaml                          sillygoose@Bryces-MacBook-Pro
apiVersion: karpenter.sh/v1beta1
kind: NodePool
.....
spec:
  disruption:
    budgets:
    - nodes: 10%
    consolidationPolicy: WhenUnderutilized

```

> Lets choose total number of nodes, since this allows the user to also specify periods where no nodes are to be disrupted of a particular type of disruption, and makes more sense with the existing karpenter behavior today.

Reviewer 3: I think you're proposing that everything but (drift, underutilized, empty) should be unbounded in disruption, is that right?

Author: Yes that is correct. Any disruption that falls under a different reason, when there is no default semantic (Empty reason with nodes specified) will be unbounded in that case.

# API Design
## Approach A: Extending the Budget API to specify a reason 
This section contrasts two approaches for specifying disruption reasons in the v1beta1 nodepool API: a list of reasons (List Approach) and a single reason per budget (Single Reason Approach).

> ### List Approach: Multiple Reasons per Budget

Reviewer 3: Worth putting something like this to guide the users that this is what you're thinking of.
```suggestion
### List Approach: Multiple Reasons per Budget - Recommended
```

This approach allows specifying multiple disruption methods within a single budget entry. It is proposed to add a field Reasons to the budgets, which can include a list of reasons this budget applies to.
#### Proposed Spec
Add a simple field "reasons" is proposed to be added to the budgets. 
```go
// Budget defines when Karpenter will restrict the
// number of Node Claims that can be terminating simultaneously.
type Budget struct {
      // Reasons is a list of methods for disruption that apply to this budget. If Reasons is not set, this budget applies to all methods.
      // If a reason is set, it will only apply to that method. If multiple reasons are specified,
      // this budget will apply to all of them. If a budget is not specified for a method, the default budget will be used.

>       // allowed reasons are "consolidation", "expired", "empty", "drift"

Reviewer 4: ```suggestion
      // allowed reasons are "underutilized", "expired", "empty", "drift"
```

Consolidation _has_ a policy that covers empty and underutilized so I think that we should rename this from underutilized to empty. Realistically, this is the difference between the number of nodes that we can disrupt that aren't fully empty vs. ones that are, which I think is what users really care about here.

Reviewer 4: The other question that I have here: do we need something that describes the difference between replace and delete? Maybe this is an extension into the future, but I could imagine _this_ is what users care about here too. e.g. When you disrupt my node, is it going to go onto existing capacity (e.g. I can trust that it will run because pods are already there) or is it going to go onto new capacity? 

The reason that I call this out is that when we are doing a consolidation replace for a NodePool that has drifted configuration, the replacement node is still going to have the new configuration, even if we have fully blocked drift, which may be what the user was trying to avoid in the first place. Have we thought about this case and if this is what users want/expect?

>       // If a reason == "default", it will apply to all reasons that don't have an active budget. If a reason is set, it will only apply to that method. If multiple reasons are specified,
>       // this budget will apply to all of them. If a budget is not specified for a method, the default budget will be used.
>       // allowed reasons are "default", "consolidation", "expired", "empty", "drift"

Reviewer 3: I think we discussed that default == omitting this field, right? Can we update this description to match that? Looks like the below kubebuilder validation reflects this. 
```suggestion
      // If a reason is set, it will only apply to that method. If multiple reasons are specified,
      // this budget will apply to all of them. If a budget is not specified for a method, the default budget will be used.
      // allowed reasons are "consolidation", "expired", "empty", "drift"
```

Reviewer 3: Also what does a "default budget" mean here? Can we just say it's unbounded?

      // +kubebuilder:validation:MaxItems=5

>       // +kubebuilder:validation:Enum:={"consolidation","expired","empty","drift"}

Reviewer 4: nit: Are these reasons capitalized or lower-case? I typically see policies and reasons in Kubernetes APIs in upper-case, so I would tend to lean towards upper-case for these

Author: Good point. Uppercase makes more sense. Initially they moved to lowercase when I was using the Method.Type() to reference each of the reasons. 

Now that those semantics have been axed in favor of expired, drifted, underutilized, etc we can capitalize these.

If there are multiple active budgets, karpenter takes the most restrictive budget. This same principle will be applied to the disruption budgets in this approach. The only difference in behavior is that each window will apply to list of reasons that are specifed rather than just all disruption methods. 
### Pros + Cons 
👍👍 Flexibility in Budget Allocation: Allows more flexibility in allocating budgets across multiple disruption reasons.
👍👍 Reduced Configuration Complexity: Simplifies the configuration process, especially for similar settings across multiple reasons.
👎 Potential for API Complexity: There might be confusion over whether the node count is shared between actions or if each action gets the node count individually. 

Note some pros and cons between A + B can be shared, and are in a list next to the pros + cons for Single Reason Approach 

#### Single Reason Approach: One Reason per Budget

### Proposed Spec
In this approach, each budget entry specifies a single reason for disruption.

```go
// Budget defines when Karpenter will restrict the
// number of Node Claims that can be terminating simultaneously.
type Budget struct {

>       // +kubebuilder:validation:Enum:={"all", "default","consolidation","expired","empty","drift"}

Reviewer 3: I think  all and default are from a previous iteration, lmk if i'm wrong here. 
```suggestion
      // +kubebuilder:validation:Enum:={"consolidation","expired","empty","drift"}
```

      // +optional
      Reason string `json:"reason,omitempty" hash:"ignore"`
      // Nodes dictates the maximum number of NodeClaims owned by this NodePool
      // that can be terminating at once. This is calculated by counting nodes that
      // have a deletion timestamp set, or are actively being deleted by Karpenter.
      // This field is required when specifying a budget.
      // This cannot be of type intstr.IntOrString since kubebuilder doesn't support pattern
      // checking for int nodes for IntOrString nodes.
      // Ref: https://github.com/kubernetes-sigs/controller-tools/blob/55efe4be40394a288216dab63156b0a64fb82929/pkg/crd/markers/validation.go#L379-L388
      // +kubebuilder:validation:Pattern:="^((100|[0-9]{1,2})%|[0-9]+)$"
      // +kubebuilder:default:="10%"
      Nodes string `json:"nodes" hash:"ignore"`
      // Schedule specifies when a budget begins being active, following
      // the upstream cronjob syntax. If omitted, the budget is always active.

>       // Timezones are not supported.

Reviewer 6: I can already taste the sadness of some poor American engineer having to update the disruption policies twice a year for DST.

      // +optional
      Schedule *string `json:"schedule,omitempty" hash:"ignore"`
      // Duration determines how long a Budget is active since each Schedule hit.
      // Only minutes and hours are accepted, as cron does not work in seconds.
      // If omitted, the budget is always active.
      // This is required if Schedule is set.
      // This regex has an optional 0s at the end since the duration.String() always adds
      // a 0s at the end.
      // +kubebuilder:validation:Pattern=`^([0-9]+(m|h)+(0s)?)$`
      // +kubebuilder:validation:Type="string"
      // +optional
      Duration *metav1.Duration `json:"duration,omitempty" hash:"ignore"`
}

```
##### Example
```yaml
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: default
spec: # This is not a complete NodePool Spec.
  disruption:
    budgets:
    - schedule: "* * * * *"
      reason: drifted 
      nodes: 10
    # For all other reasons , only allow 5 nodes to be disrupted at a time
    - nodes: 5
      schedule: "* * * * *"

```

#### Pros and Cons
👍 Simplicity and Clarity: Offers a straightforward and clear mapping of budget to disruption reason.

> 👎 Increased Configuration Overhead: Requires duplicating settings for multiple reasons, increasing setup complexity.

Reviewer 3: Making it easier to reach the 50 limit as well.

Reviewer 4: To be fair, we could always bump the limit. Having multiple reasons allows users to read their budgets a little easier. If we expect users to lump expiration and drift together, I feel like it makes sense to make this a list.

### Approach: Defining Per Reason Controls  
Ideally, we could move all generic controls that easily map into other reasons into one set of reason controls, this applies to budgets and other various disruption controls that could be more generic. 
#### Example 

```yaml 
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: example-nodepool
spec:
  disruption:
    defaults:
      budgets: 
        - nodes: 10% 
          schedule: "0 0 1 * *"
          duration: 1h 
    consolidation:
      consolidationPolicy: WhenUnderutilized
      disruptAfter: "30m"
    drift:
      budgets:
        - nodes: "20%"
          schedule: "0 0 1 * *"
          duration: "1h"
        - nodes: "10%"
          schedule: "0 0 * * 0"
          duration: "2h"
        - nodes: "50%" 
          schedule: "@monthly"
```
#### Considerations 

> Some of the API choices for a given reason seem to follow a similar pattern. These include ConsolidateAfter, ExpireAfter, and there are discussions about introducing a global DisruptAfter. Moreover, when discussing disruption budgets, we talk about adding behavior for each reason. It appears there is a need for disruption controls within the budgets for each reason, not just overall.

Reviewer 3: May not make as much sense if we move the expireAfter into the NodeClaimTemplate itself, but considering this isn't the recommended option, I think we're good here.

Author: Agreed.

This approach aligns well with controls that apply to all existing reasons. The proposal presented here is similar to the one mentioned above in relation to the reasons we allow to be defined (underutilized, drifted, expired, empty).

This proposal is currently scoped for disruptionBudgets by reason. However, we should also consider incorporating other generic disruption controls into the PerReasonControls, even if we do not implement them immediately. Moving ConsolidateAfter and ExpireAfter into the per-reason controls is a significant migration that requires careful planning and its own dedicated design. This proposal simply demonstrates a potential model that highlights the benefits of defining controls at a per-reason level of granularity.

> ### Pros + Cons 

Reviewer 4: The other problem that I have with this approach is that I think that this tries to group things together too early without having strong reasoning from users that we should do it this way. We actually got burned by this by grouping all of the disruption behavior into the same stanza. We're now thinking that we should undo some of this, since it's looking like some of this behavior should get pushed down into the NodeClaim API now

Author: I agree that defining PerReasonControls before we have a large set of controls needed for each reason is too aggressive and one of the main reasons I axed this approach.

### Q: How should allowed disruptions be calculated in a multi-reason world? 
The calculation of allowed disruptions in a system with multiple disruption reasons has become more intricate following the introduction of Disruption Budget Reasons. Previously, the formula was straightforward:

AllowedDisruptions = minNodeCountOfAnActiveBudget - unhealthyNodes - totalDisruptingNodes.

When calculating by reason, two potential equations emerge:
1. AllowedDisruptionByReason = minNodeCountOfActiveBudget[reason] - unhealthyNodes - totalDisruptingNodes[reason]
2. AllowedDisruptionByReason = minNodeCountOfActiveBudget[reason] - unhealthyNodes - totalDisruptingNodes
The second equation is the one we should opt into.

Take this budget as an example
```yaml
budgets: 
  - nodes: 15 
    reasons: [Drift, Consolidation]
    schedule: "* * * * *"
  - nodes: 10
    reasons: [Drift]
    schedule: "* * * * *"
  - nodes: 30 
    schedule: "* * * * *" 
```
First, calculate minNodeCountOfActiveBudget. Given there are two budgets for Drift, we select the lower number:

```
minNodeCountOfActiveBudget = {
  Drift: 10 
  Consolidation: 15 

>   Default: 5

Author: 30 is supposed to be 5.

    schedule: "* * * * *"
  - nodes: 10
    reasons: [Drift]
    schedule: "* * * * *"
  - nodes: 30 
    schedule: "* * * * *" 
```

First we need to calculate the minNodeCountOfActiveBudget, in this case we note that we set 10 for Drift, since there are two active budgets for drift we take the minimum.
minNodeCountOfActiveBudget = {
  Drift: 10 
  Consolidation: 15 
  Default: 5
}

If we use the first equation, we then need to also track in karpenter state the number of nodes currently being disrupted by a given disruption action. 
disrupting = {
  Drift: 3 
  Consolidation: 6
  Expriation: 3
  Emptiness: 2
}

Let say unhealthy nodes are zero. 

**First Equation**
Drift = 10 - 3 # We can take disruption, because the Drift Bucket isn't full
Consolidation = 15 - 6 # We can disrupt because 15 - 6 > 0
Default = 5 - Expriation - Emptiness = 0, # We cannot disrupt any other reasons than consolidation and drift
**Second Equation** 
totalDisrupting = 3 + 6 + 5 = 14
Drift = 10 - 14 # We cannot disrupt due to drift 
Consolidation = 15 - 14 # We can disrupt due to consolidation at leasrt for one node.
Default: 5 - 14 # We cannot disrupt any other methods

> With the second equation, its much simpler to reason about. A node can be marked for multiple disruption reasons(Emptiness, Drift, Consolidation). SO its challenging to calculate in flight, which exact disruption you will end up hitting. With the second equation, it makes reasoning about how many nodes are getting disrupted simpler for cluster operators.

Author: @njtran and I talked about this second equation, and came to the conclusion its the simpler behavior to understand. 

When you have 

N Nodepools 
50 Budgets, 

Tracking these behaviors is too complex for users to properly reason about the buckets. 

@garvinp-stripe  what are your thoughts about rather than bucketing by reason, we instead respect the total disrupting count instead? Also makes it easier to address the edge case you and I spoke about

Reviewer 7: At first glance I think that make sense especially since the outcome is more conservative (less nodes disrupted) and untangles the messy logic when nodes are disrupted by multiple reasons. From my perspective this is good. Will let you know if any other thoughts comes up.

Reviewer 5: I like the second equation more as well.

Author: ok everyone agreees then we go with second equation.

>  Disruption Controls By Reason

Reviewer 3: I think you probably want to keep the top level title
```suggestion
# Disruption Controls By Reason
```

    budgets:
    - schedule: "* * * * *"
      reason: drifted 
      nodes: 10
    # For all other reasons , only allow 5 nodes to be disrupted at a time
    - nodes: 5
      schedule: "* * * * *"

```

#### Pros and Cons
* 👍 Simplicity and Clarity: Offers a straightforward and clear mapping of budget to disruption reason.
* 👎 Increased Configuration Overhead: Requires duplicating settings for multiple reasons, increasing setup complexity.
* 👎 Less Flexibility: Lacks the flexibility to share a budget across multiple reasons.

#### Pros and Cons for List and Per Reason Definitions
Some of the Pros and Cons are shared by both list and single reason, as they have the same advantages and disadvantages in comparison to Per Reason Controls 
* 👍👍 Extends Existing API:  No Breaking API Changes, completely backwards compatible
* 👍 No Nesting Required: Leaves budgets at the top level of the api.
* 👎 Limited Generalization of Reason Controls: With reason being clearly tied to budgets, and other api logic being driven by disruption reason, we lose the chance to generalize per Reason controls. If we ever decide we need a place per action,  there will be some duplication for reason. 

> ### Preferred Option: List Approach

Reviewer 4: Should the preferred option be at the very end? Or you are claiming that this is the preferred approach of the extend budgets with reasons options?

Author: Its the preferred option overall, I can move it to the top.

Author: Wait went and reread my document sorry tis been a while. Thats there to show the prefered option between the two. 

The list option is still preferred overall, but that is the preferred choice between the two options(single reason and multiple reason).

Author: Ok broke them up structurally, but I did leave one bit tieing the two together as I do feel strongly the two are related.

Given the comparison, the preferred design is the List Approach. It provides the necessary flexibility for managing multiple disruption reasons under a single budget, while reducing configuration complexity. This approach extends the existing API without introducing breaking changes and simplifies management for scenarios where multiple disruption reasons share similar constraints.

> ### Approach: Defining Per Reason Controls  

Reviewer 4: ```suggestion
### Approach B: Defining Per Reason Controls  
```

### Pros + Cons 
* 👍 Granular Control for Each Reason: This model aligns well with the need for specific controls for each disruption reason, providing a natural specification for further granularity. 
* 👍 Foundation for Future Extensions: Offers a framework for incorporating other generic disruption controls into the per-reason controls, extending beyond budgets.
* 👍👍 Customization for Specific Reasons: Allows for specific validations and extensions for individual reasons, offering tailored control.
* 👎👎👎 Breaking API Changes:Implementing this approach would necessitate significant changes to the current API, especially for budgets. If we decide to model DisruptAfter in a similar way we also would have to break those apis. It might make sense to break the budgets now before they have garnered large adoption as it becomes harder to make the change as time goes on.
* 👎 Complexity in Default Reason Handling: The model does not facilitate easy defaulting for unspecified ("default") disruption reasons.
* 👎 Increased Budget Complexity: The use of budgets becomes more complex as they are now nested within another field, adding a layer of intricacy to their application.

> ### API Design Conclusion: Preferred Design

Reviewer 4: Should this just directly reference approach A?

Author: It does now

See Less Made Up Scenarios here: 
- https://github.com/kubernetes-sigs/karpenter/issues/924 
- https://github.com/kubernetes-sigs/karpenter/issues/753#issuecomment-1790110838
- https://github.com/kubernetes-sigs/karpenter/issues/672

See further breakdown of some additional scenarios towards the bottom of the document 

## Clarifying the requirements and behavior 
**Reason and Budget Definition:** Users should be able to define an reason and a corresponding budget(s).
**Supported Reasons:** All disruption Reasons affected by the current Budgets implementation (underutilized, empty, expired, drifted) should be supported. 
**Default Behavior for Unspecified Reasons:** Budgets should continue to support a default behavior for all disruption reasons. 

# API Design
## Approach A: Extending the Budget API to specify a reason 
This section contrasts two approaches for specifying disruption reasons in the v1beta1 nodepool API: a list of reasons (List Approach) and a single reason per budget (Single Reason Approach).

> ### List Approach: Multiple Reasons per Budget - Recommended

Reviewer 8: @Bryce-Soghigian as this solution is recommended I would like to ask if we can extend this approach with `spec.disruption.defaults.budgets` section as proposed in `Approach: Defining Per Reason Controls`.
The idea is to support use case from [this](https://github.com/kubernetes-sigs/karpenter/discussions/1079#discussioncomment-8719765) discussion.
In short:
- by default -> do not tolerate any disruption (the default one)
- in specified window -> allow 25% disruption due to drift reason
Use case: we want to keep our infrastructure intact, with an exception when we want to update the fleet during the maintenance window (which happens at weekends).

Reviewer 8: ```yaml
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: default
spec: # This is not a complete NodePool Spec.
  disruption:
    defaults:
      budgets: 
        - nodes: 0% 
          schedule: "* * * * *"
    budgets:
    - schedule: "0 17 * * sat"
      reasons: [drifted, underutilized]
      nodes: 25%
```

Reviewer 3: So you'd want to explicitly define the "If no other budgets are active, use this budget" type of behavior? This sounds like it'd be better specified through a different feature request. Are you willing to open up an issue for this @phajduk ?

Author: @phajduk that should be a separate RFC. I have been thinking about that kind of behavior but it doesn't belong in the per reason controls design.

}

```
#### Example
```yaml
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: default
spec: # This is not a complete NodePool Spec.
  disruption:
    budgets:
    - schedule: "* * * * *"
      reasons: [drifted, underutilized]
      nodes: 10
    # For all other reasons, only allow 5 nodes to be disrupted at a time
    - nodes: 5
      schedule: "* * * * *"

```

In the original proposed spec, karpenter allows the user to specify up to [50 budgets](https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/apis/v1beta1/nodepool.go#L96)
If there are multiple active budgets, karpenter takes the most restrictive budget. This same principle will be applied to the disruption budgets in this approach. The only difference in behavior is that each window will apply to list of reasons that are specified rather than just all disruption methods. 
### Pros + Cons 
* 👍👍 Flexibility in Budget Allocation: Allows more flexibility in allocating budgets across multiple disruption reasons.
* 👍👍 Reduced Configuration Complexity: Simplifies the configuration process, especially for similar settings across multiple reasons.

> * 👎 Potential for API Complexity: There might be confusion over whether the node count is shared between actions or if each action gets the node count individually. 

Author: We talk more about this later

      // +optional
      Duration *metav1.Duration `json:"duration,omitempty" hash:"ignore"`
}

```
##### Example
```yaml
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: default
spec: # This is not a complete NodePool Spec.
  disruption:
    budgets:
    - schedule: "* * * * *"
      reason: drifted 
      nodes: 10
    # For all other reasons , only allow 5 nodes to be disrupted at a time
    - nodes: 5
      schedule: "* * * * *"

```

#### Pros and Cons
* 👍 Simplicity and Clarity: Offers a straightforward and clear mapping of budget to disruption reason.
* 👎 Increased Configuration Overhead: Requires duplicating settings for multiple reasons, increasing setup complexity.
* 👎 Less Flexibility: Lacks the flexibility to share a budget across multiple reasons.

#### Pros and Cons for List and Per Reason Definitions
Some of the Pros and Cons are shared by both list and single reason, as they have the same advantages and disadvantages in comparison to Per Reason Controls 
* 👍👍 Extends Existing API:  No Breaking API Changes, completely backwards compatible
* 👍 No Nesting Required: Leaves budgets at the top level of the api.

> * 👎 Limited Generalization of Reason Controls: With reason being clearly tied to budgets, and other api logic being driven by disruption reason, we lose the chance to generalize per Reason controls. If we ever decide we need a place per action,  there will be some duplication for reason. 

Author: This point is more so for future controls that may surface in the API. For example, we could have other controls we have been discussing like MinAge apply per disruption reason. 

 I may only want to disrupt for cost savings after my nodeclaim has lived for a certain amount of time. I don't care that you remove empty nodes right away. On the other hand, nodes that are just underutilized, I want to only have karpenter remove them when their Age is greater than 10m.  Don't disrupt active workloads right away.
 
 MinAge, or DisruptAfter are both concepts we have talked about a bit that might make sense to have it be scoped per disruption reason as well. Rather than ConsolidateAfter, we could have a more generic disruptAfter. 
 
 
 Overall we decided against "Approach C" which involves defining more generic per reason groups due to the large API surface area that would have to change. It is just worth mentioning that we lose the ability to group disruption behaviors or API surface by reason in this design.

          schedule: "@monthly"
```
#### Considerations 
Some of the API choices for a given reason seem to follow a similar pattern. These include ConsolidateAfter, ExpireAfter, and there are discussions about introducing a global DisruptAfter. Moreover, when discussing disruption budgets, we talk about adding behavior for each reason. It appears there is a need for disruption controls within the budgets for each reason, not just overall.

This approach aligns well with controls that apply to all existing reasons. The proposal presented here is similar to the one mentioned above in relation to the reasons we allow to be defined (underutilized, drifted, expired, empty).

This proposal is currently scoped for disruptionBudgets by reason. However, we should also consider incorporating other generic disruption controls into the PerReasonControls, even if we do not implement them immediately. Moving ConsolidateAfter and ExpireAfter into the per-reason controls is a significant migration that requires careful planning and its own dedicated design. This proposal simply demonstrates a potential model that highlights the benefits of defining controls at a per-reason level of granularity.

> ### Pros + Cons 

Reviewer 9: Another way we could factor the API to provide the kind of composability I think you're looking for with this approach is via a separate CRD. Although, I don't think this is necessary to meet the specific needs of these use cases (or any that I have line of sight to in the future)

* 👍 Foundation for Future Extensions: Offers a framework for incorporating other generic disruption controls into the per-reason controls, extending beyond budgets.
* 👍👍 Customization for Specific Reasons: Allows for specific validations and extensions for individual reasons, offering tailored control.
* 👎👎👎 Breaking API Changes:Implementing this approach would necessitate significant changes to the current API, especially for budgets. If we decide to model DisruptAfter in a similar way we also would have to break those apis. It might make sense to break the budgets now before they have garnered large adoption as it becomes harder to make the change as time goes on.
* 👎 Complexity in Default Reason Handling: The model does not facilitate easy defaulting for unspecified ("default") disruption reasons.
* 👎 Increased Budget Complexity: The use of budgets becomes more complex as they are now nested within another field, adding a layer of intricacy to their application.

### API Design Conclusion: Preferred Design

> After evaluating different approaches to extend the Karpenter API for specifying disruption reasons, the preferred design is the List Approach in Approach A. This approach offers flexibility in managing multiple disruption reasons under a single budget and reduces configuration complexity. It extends the existing API without introducing breaking changes and simplifies management for scenarios where multiple disruption reasons share similar constraints.

Reviewer 9: I'm aligned with this approach

While the idea of per-reason controls (Approach B) provides granular control and a foundation for future extensions, it involves significant API changes and increased complexity, making it less favorable at this stage. However, this approach remains a viable option for future considerations, especially if there is a need for more tailored control over each disruption reason.

### Disruption Behavior
The calculation of allowed disruptions in a system with multiple disruption reasons has become more intricate following the introduction of Disruption Budget Reasons. Previously, the formula was straightforward:

> AllowedDisruptions = minNodeCount - unhealthyNodes - totalDisruptingNodes.

Reviewer 9: What is `minNodeCount`?

Author: min node count is the minimum specifed node count in the crd. 

For budgets, I specify 
- nodes: 5
- nodes: 10 

min node count for these budgets would be 5.

> When calculating by reason, three potential equations emerge:

Reviewer 9: It wasn't clear to me which of these we ended up recommending. Can you add a callout below like in the above decisions?

1. BucketedDisruptionByReason = minNodeCount[reason] - unhealthyNodes - totalDisruptingNodes[reason]
2. AllowedDisruptionByReason = minNodeCount[reason] - unhealthyNodes - totalDisruptingNodes

> 3. minimumDisruptionAllowed = min(minNodeCount[reason], minNodeCount[default]) - unhealthyNodes - totalDisruptingNodes

Reviewer 9: I think this equation is the scenario I was asking about earlier

  disruption:
    consolidationPolicy: WhenEmpty 
    expireAfter: 7d 
    budgets:
    - nodes: "100%" # Always allow Expiration and Emptiness to disrupt 
      reasons: [expired, empty]
    - nodes: 0
```
So in this case, we would take the minimum node count, and it would block the 100% expiration budget of expired and Empty. 
But this semantic would not work for equation 3. Equation three will take the minimum between minNodeCount[reason] and the minNodeCount[default/unspecifed].  

#### Fully Blocking Drift && Expiration while allowing consolidation
```yaml 
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: worker-pool 
spec: # This is not a complete NodePool Spec.
  disruption:
    consolidationPolicy: WhenUnderutilized 
    expireAfter: Never 
    budgets:
    - nodes: 0
      reasons: [drifted, expired] 
    - nodes: "100%" // All Disruption Other than drift and expiration are allowed
```

#### Fully Blocking Drift and Expiration while allowing emptiness 
```yaml 
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: worker-pool 
spec: # This is not a complete NodePool Spec.
  disruption:
    consolidationPolicy: WhenEmpty

>     expireAfter: Never 

Reviewer 9: The interaction of `expireAfter: Never` and `budgets: [nodes: 0, reasons: [expired]]` is a bit unsatisfying. I wonder if there are edge cases that'll emerge from having two ways to accomplish the same result

Author: I agree. I used the expireAfter: Never controls here to call that out implicitly that we have multiple controls for blocking expire. The never semantic is conflicting with the budgets semantic. 

I can define 

```
disruption: 
     expireAfter: Never 
     
     budgets: 
        - nodes: 10 
          reasons: ["expired"] 
          
 ```
 
In this conflicting example, our behavior still adheres to the principle we set in the budgets. In this case we respect the lower disruption setting anyway. If i set expire never we will not expire nodes. 

I can similarly define a budget with conflicting values.  
```
disruption: 
    budgets: 
      - nodes: 10 
         reasons: ["expired"]
      - nodes: 15 
         reasons: ["expired"]
```

We take 10 rather than 15. In that sense expireAfter is disrupt 0 due to expire and the budget of 10 from the disruption budgets, we still take the disruption action that is least disruptive.

    budgets:
    - nodes: 0
      reasons: [drifted] 
    - nodes: "100%" // All Disruption Other than drift(Expiration and Empty Nodes) are allowed
```

#### Limiting Drift & Expiration while allowing aggressive but not unbounded consolidation

> Drift and Expiration may not be desired reasons to remove many nodes in one interval, but Underutilization and Emptiness may be desired. We may want to still consolidate at a healthy threshold but not allow 

Reviewer 9: not allow...?

Author: Fixed that good catch

```yaml
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: worker-pool 
spec: # This is not a complete NodePool Spec.
  disruption:
    consolidationPolicy: WhenUnderutilized 
    expireAfter: 1h 
    budgets:
    - nodes: 1
      reasons: [drifted, expired] # Only allow Drift and Expiration to occur one at a time 
    - nodes: "33%" # All Consolidation actions taken either through replace or delete should be allowed to disrupt at least a third of the cluster at once 
```

### Q: How should we handle an unspecified reason when others are specified? 
```yaml
budgets: 
  - nodes: 10
    reasons: [drifted, underutilized]
    schedule: "* * * * *"
  - nodes: 5 
    reasons: [empty] 
    schedule: "* * * * *" 
```

> In the case of a budget like above, default is undefined. Should karpenter assume the user doesn't want to disrupt any other reasons? Or should we assume that if a default is unspecified, they want us to disrupt anyway?  

Reviewer 9: How is the default undefined? I would assume the default is the `reasons: []` budget

Author: in this particular example 

```yaml
budgets: 
  - nodes: 10
    reasons: [drifted, underutilized]
    schedule: "* * * * *"
  - nodes: 5 
    reasons: [empty] 
    schedule: "* * * * *" 
```
We have a disruption budget mapping of 
budgetMapping = {
   drifted: 10 
   underutilized: 10 
   empty: 5 
}

See here we have no budget defined that covers the behavior of expired. So what should we default the values for "expired" disruption? Should we allow disruption for expired nodes? Is it unbounded since no restriction is specified?  That's what I mean. 

 The "default" language is a remnant from earlier revisions of this "default" concept I had in the original design document. In this case, I am asking what should the default value be for a reason with an undefined disruption budget.

### List Approach: Multiple Reasons per Budget - Recommended
This approach allows specifying multiple disruption methods within a single budget entry. It is proposed to add a field Reasons to the budgets, which can include a list of reasons this budget applies to.
#### Proposed Spec
Add a simple field "reasons" is proposed to be added to the budgets. 
```go
// Budget defines when Karpenter will restrict the
// number of Node Claims that can be terminating simultaneously.
type Budget struct {
      // Reasons is a list of disruption reasons. If Reasons is not set, this budget applies to all methods.
      // If a reason is set, it will only apply to that method. If multiple reasons are specified,
      // this budget will apply to all of them. If a reason is unspecified we will take the min value of this budget and the rest of the active budgets.
      // if an unspecified reason exists we will also override all other reasons with its value if they are smaller than the unspecified reason.
      // allowed reasons are "underutilized", "expired", "empty", "drifted"

>       // +kubebuilder:validation:MaxItems=5

Reviewer 3: Why would it be 5 items? Is there a way to ensure values are unique?

Author: good catch we no longer have 5 items as default is gone. This should be 4.

Author: I haven't tested this, but the docs show we can use Unique items 

// +kubebuilder:validation:UniqueItems
bool
specifies that all items in this list must be unique.

Author: See: https://book.kubebuilder.io/reference/markers/crd-validation for reference

>       // +kubebuilder:validation:Enum:={"underutilized","expired","empty","drifted"}

Reviewer 3: I think we can handle this in the feature, but we should really consider modeling this after the status conditions that we eventually put on the node claim for each of these.

Author: I agree i had a section in this doc outlining observability and extending the nodeclaim to have each of these but that got scraped. Separate work is required for observability of the budgets in general.

      // +optional
      Reasons []string `json:"reason,omitempty" hash:"ignore"`
      // Nodes dictates the maximum number of NodeClaims owned by this NodePool
      // that can be terminating at once. This is calculated by counting nodes that
      // have a deletion timestamp set, or are actively being deleted by Karpenter.
      // This field is required when specifying a budget.
      // This cannot be of type intstr.IntOrString since kubebuilder doesn't support pattern
      // checking for int nodes for IntOrString nodes.
      // Ref: https://github.com/kubernetes-sigs/controller-tools/blob/55efe4be40394a288216dab63156b0a64fb82929/pkg/crd/markers/validation.go#L379-L388
      // +kubebuilder:validation:Pattern:="^((100|[0-9]{1,2})%|[0-9]+)$"
      // +kubebuilder:default:="10%"
      Nodes string `json:"nodes" hash:"ignore"`
      // Schedule specifies when a budget begins being active, following

>       // the upstream cronjob syntax. If omitted, the budget is always active. // Timezones are not supported. // This field is required if Duration is set. // +kubebuilder:validation:Pattern:=`^(@(annually|yearly|monthly|weekly|daily|midnight|hourly))|((.+)\s(.+)\s(.+)\s(.+)\s(.+))$` // +optional Schedule *string `json:"schedule,omitempty" hash:"ignore"` // Duration determines how long a Budget is active since each Schedule hit. // Only minutes and hours are accepted, as cron does not work in seconds.

Reviewer 3: nit: fix the new lines on this
```suggestion
      // the upstream cronjob syntax. If omitted, the budget is always active. 
      // Timezones are not supported. This field is required if Duration is set. 
      // +kubebuilder:validation:Pattern:=`^(@(annually|yearly|monthly|weekly|daily|midnight|hourly))|((.+)\s(.+)\s(.+)\s(.+)\s(.+))$` 
      // +optional Schedule *string `json:"schedule,omitempty" hash:"ignore"` 
      // Duration determines how long a Budget is active since each Schedule hit. 
      // Only minutes and hours are accepted, as cron does not work in seconds.
```

This section contrasts two approaches for specifying disruption reasons in the v1beta1 nodepool API: a list of reasons (List Approach) and a single reason per budget (Single Reason Approach).

### List Approach: Multiple Reasons per Budget - Recommended
This approach allows specifying multiple disruption methods within a single budget entry. It is proposed to add a field Reasons to the budgets, which can include a list of reasons this budget applies to.
#### Proposed Spec
Add a simple field "reasons" is proposed to be added to the budgets. 
```go
// Budget defines when Karpenter will restrict the
// number of Node Claims that can be terminating simultaneously.
type Budget struct {
      // Reasons is a list of disruption reasons. If Reasons is not set, this budget applies to all methods.
      // If a reason is set, it will only apply to that method. If multiple reasons are specified,
      // this budget will apply to all of them. If a reason is unspecified we will take the min value of this budget and the rest of the active budgets.
      // if an unspecified reason exists we will also override all other reasons with its value if they are smaller than the unspecified reason.

>       // allowed reasons are "underutilized", "expired", "empty", "drifted"

Author: ```suggestion
      // allowed reasons are "underutilized", "expired", "empty", "drifted"
      // +kubebuilder:validation:UniqueItems
```

>       // +kubebuilder:validation:MaxItems=5

Author: ```suggestion
      // +kubebuilder:validation:MaxItems=4
```

# Disruption Controls By Reason

> # Table of Contents

Reviewer 4: Does using `[[_TOC_]]` work so we don't have to make sure this constantly aligns with the naming/titles below? See: https://stackoverflow.com/questions/47154661/how-can-i-create-a-table-of-content-in-gitlab-wiki

Author: https://github.blog/changelog/2021-04-13-table-of-contents-support-in-markdown-files/ from what I can see github doesn't natively support generating them and trying the `[[_TOC_]]` syntax but it did not generate what we are looking for. 

For now, we can remove the TOC, or I can regenerate it with a separate tool once we are ready to merge this in.

1. [Disruption Controls By Reason](#disruption-controls-by-reason)
2. [API Design](#api-design)
   1. [Approach A: Extending the Budget API to Specify a Reason](#approach-a-extending-the-budget-api-to-specify-a-reason)
      1. [List Approach: Multiple Reasons per Budget - Recommended](#list-approach-multiple-reasons-per-budget---recommended)
      2. [Single Reason Approach: One Reason per Budget](#single-reason-approach-one-reason-per-budget)
   2. [Approach B: Defining Per Reason Controls](#approach-b-defining-per-reason-controls)

>    3. [Preferred Option: List Approach](#preferred-option-list-approach)

Reviewer 4: Can we make sure that this ToC matches up with the actual structure of headers in the document? Looking through the doc, I was noticing that some things that should have been "h2" were listed as "h3" and vice vresa.

Author: Talked about elsewhere

   4. [Disruption Behavior](#disruption-behavior)

>    5. [Scenarios](#scenarios)

Reviewer 4: ```suggestion
   5. [User Stories](#scenarios)
```

This is a nit, but most KEPs call them user stories, make sure the header below is aligned with this naming

Author: I was wondering about this. Should we be using the KEP template now that we are part of the kubernetes project proper?

* 👎 Complexity in Default Reason Handling: The model does not facilitate easy defaulting for unspecified ("default") disruption reasons.
* 👎 Increased Budget Complexity: The use of budgets becomes more complex as they are now nested within another field, adding a layer of intricacy to their application.

### API Design Conclusion: Preferred Design
After evaluating different approaches to extend the Karpenter API for specifying disruption reasons, the preferred design is the List Approach in Approach A. This approach offers flexibility in managing multiple disruption reasons under a single budget and reduces configuration complexity. It extends the existing API without introducing breaking changes and simplifies management for scenarios where multiple disruption reasons share similar constraints.

While the idea of per-reason controls (Approach B) provides granular control and a foundation for future extensions, it involves significant API changes and increased complexity, making it less favorable at this stage. However, this approach remains a viable option for future considerations, especially if there is a need for more tailored control over each disruption reason.

> ### Disruption Behavior

Reviewer 4: I wonder if we could use something more descriptive for this title like, "Counting Disruptions Against a Budget" or something like this that's more tailored to the question we are asking here

The calculation of allowed disruptions in a system with multiple disruption reasons has become more intricate following the introduction of Disruption Budget Reasons. Previously, the formula was straightforward:

AllowedDisruptions = minNodeCount - unhealthyNodes - totalDisruptingNodes.

When calculating by reason, three potential equations emerge:
1. BucketedDisruptionByReason = minNodeCount[reason] - unhealthyNodes - totalDisruptingNodes[reason]
2. AllowedDisruptionByReason = minNodeCount[reason] - unhealthyNodes - totalDisruptingNodes

> 3.(Recommended) minimumDisruptionAllowed = min(minNodeCount[reason], minNodeCount[default]) - unhealthyNodes - totalDisruptingNodes

Reviewer 4: Should the recommended one be first?

Author: Changed the ordering

**Reason and Budget Definition:** Users should be able to define an reason and a corresponding budget(s).
**Supported Reasons:** All disruption Reasons affected by the current Budgets implementation (underutilized, empty, expired, drifted) should be supported. 
**Default Behavior for Unspecified Reasons:** Budgets should continue to support a default behavior for all disruption reasons. 

# API Design
## Approach A: Extending the Budget API to specify a reason 
This section contrasts two approaches for specifying disruption reasons in the v1beta1 nodepool API: a list of reasons (List Approach) and a single reason per budget (Single Reason Approach).

### List Approach: Multiple Reasons per Budget - Recommended
This approach allows specifying multiple disruption methods within a single budget entry. It is proposed to add a field Reasons to the budgets, which can include a list of reasons this budget applies to.
#### Proposed Spec
Add a simple field "reasons" is proposed to be added to the budgets. 
```go
// Budget defines when Karpenter will restrict the
// number of Node Claims that can be terminating simultaneously.
type Budget struct {

>       // Reasons is a list of disruption reasons. If Reasons is not set, this budget applies to all methods.

Reviewer 5: > If Reasons is not set, this budget applies to all methods.

If equation 2 is chosen this should probably be `If Reasons is not set, this budget applies to all methods, which are not overridden.`?

      // the upstream cronjob syntax. If omitted, the budget is always active. // Timezones are not supported. // This field is required if Duration is set. // +kubebuilder:validation:Pattern:=`^(@(annually|yearly|monthly|weekly|daily|midnight|hourly))|((.+)\s(.+)\s(.+)\s(.+)\s(.+))$` // +optional Schedule *string `json:"schedule,omitempty" hash:"ignore"` // Duration determines how long a Budget is active since each Schedule hit. // Only minutes and hours are accepted, as cron does not work in seconds.
      // If omitted, the budget is always active.
      // This is required if Schedule is set.
      // This regex has an optional 0s at the end since the duration.String() always adds
      // a 0s at the end.
      // +kubebuilder:validation:Pattern=`^([0-9]+(m|h)+(0s)?)$`
      // +kubebuilder:validation:Type="string"
      // +optional
      Duration *metav1.Duration `json:"duration,omitempty" hash:"ignore"`
}

```
#### Example
```yaml
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: default
spec: # This is not a complete NodePool Spec.
  disruption:
    budgets:
    - schedule: "* * * * *"
      reason: [Drifted, Underutilized]
      nodes: 10

>     # For all other reasons , only allow 5 nodes to be disrupted at a time

Reviewer 5: hyper-nit: typo `reasons, only` (remove space before comma)

    - nodes: 5
      schedule: "* * * * *"

```

In the original proposed spec, karpenter allows the user to specify up to [50 budgets](https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/apis/v1beta1/nodepool.go#L96)
If there are multiple active budgets, karpenter takes the most restrictive budget. This same principle will be applied to the disruption budgets in this approach. The only difference in behavior is that each window will apply to list of reasons that are specifed rather than just all disruption methods. 
### Pros + Cons 

> 👍👍 Flexibility in Budget Allocation: Allows more flexibility in allocating budgets across multiple disruption reasons.

Reviewer 5: fyi: smashed all on the same line:
![image](https://github.com/kubernetes-sigs/karpenter/assets/33269602/119157c9-740b-405a-ab97-7d22ae9a875e)

Author: fixed

While the idea of per-reason controls (Approach B) provides granular control and a foundation for future extensions, it involves significant API changes and increased complexity, making it less favorable at this stage. However, this approach remains a viable option for future considerations, especially if there is a need for more tailored control over each disruption reason.

### Disruption Behavior
The calculation of allowed disruptions in a system with multiple disruption reasons has become more intricate following the introduction of Disruption Budget Reasons. Previously, the formula was straightforward:

AllowedDisruptions = minNodeCount - unhealthyNodes - totalDisruptingNodes.

When calculating by reason, three potential equations emerge:
1. BucketedDisruptionByReason = minNodeCount[reason] - unhealthyNodes - totalDisruptingNodes[reason]
2. AllowedDisruptionByReason = minNodeCount[reason] - unhealthyNodes - totalDisruptingNodes

> 3. minimumDisruptionAllowed = min(minNodeCount[reason], minNodeCount[default]) - unhealthyNodes - totalDisruptingNodes

Author: Is there a place for a max semantic to apply to a group of budgets rather than min? Some issues seem to show that customers are looking for a disruption override. For one period I define, i want to allow disruption. Thats the persona I have been seeing in a couple of places.  Lets say I always tolerate emptiness at maximum, but don't want to tolerate cost savings. 

With equation 3, this tends to miss that mark. We could potentially fix this with allowing the user to specify if this function is min(minNodeCount[reason], minNodeCount[default]) or max()

apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: default
spec: # This is not a complete NodePool Spec.
  disruption:
     budgetGreed(better name pending): Min || Max
    budgets:
    - schedule: "* * * * *"
      reasons: [emptiness]
      nodes: 100%
    - nodes: 5
      schedule: "* * * * *"

With current poc implementation we agreed on is that this would yield 5 nodes  for everything. It would be nice to say "always allow 100%" emptiness".  Wondering if letting users choose if they want they want to use max or min makes sense?

I am always ok with you removing unused nodes, but not ok with drift, and expiration and utilization occurring in large batches. (Although drift passes one candidate?)

You can still define it as 
nodes: 5 
reasons: ["drifted", "underutilized", "expired"] 

nodes: 100% 

But seems the real intention here is to say I want to allow 100% for empty and don't tolerate much disruption concurrently for other actions.

### Scenarios
We will go through scenarios that users are expecting to face to drive home if the api can satisfy the users requirements for the end users of these features
- Block Drift, but allow Expiration
- Fully Block Drift && Expiration, but allow consolidation to occur 
- Block Drift && Expiration, but allow emptiness 

> - Limit Drift && Expirationn while allowing agressive but not unbounded consolidationn

Reviewer 5: typo: `consolidation`

Author: for that note seems Expiration is not spelled as Expirationn

    consolidationPolicy: WhenEmpty 
    expireAfter: 7d 
    budgets:
    - nodes: 0
      reasons: [Drifted] 
    - nodes: "100%" // All Disruption Other than drift(Expiration and Empty Nodes) are allowed
```

We could also allow a semantic like this
```yaml
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: worker-pool 
spec: # This is not a complete NodePool Spec.
  disruption:
    consolidationPolicy: WhenEmpty 
    expireAfter: 7d 
    budgets:
    - nodes: "100%" # Always allow Expiration and Emptiness to disrupt 
      reasons: [Expired, Empty]
    - nodes: 0
```
So in this case, we would take the minimum node count, and it would block the 100% expiration budget of Expired and Empty. 
But this semantic would not work for equation 3. Equation three will take the minimum between minNodeCount[reason] and the minNodeCount[default/unspecifed].  

#### Fully Blocking Drift && Expiration while allowing consolidation
```yaml 
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: worker-pool 
spec: # This is not a complete NodePool Spec.
  disruption:
    consolidationPolicy: WhenUnderutilized 
    expireAfter: Never 
    budgets:
    - nodes: 0

>       reasons: [Drifted] 

Reviewer 5: nit: need to add `Expired` to the array too?

Reviewer 5: Ah, nvm I see:
![image](https://github.com/kubernetes-sigs/karpenter/assets/33269602/c00c87a8-2cc1-4ece-a3bf-1549c971b905)

kind: NodePool
metadata:
  name: worker-pool 
spec: # This is not a complete NodePool Spec.
  disruption:
    consolidationPolicy: WhenUnderutilized 
    expireAfter: 720h 
    budgets:
    - nodes: "33%"
      reasons: [Drifted, Expiration] # Allow Large Drift and Expiration at All Times as these are required for cluster health 
    - nodes: "0" # All Consolidation actions taken either through replace or delete should be allowed to disrupt at least a third of the cluster at once 
      schedule: "* * * * mon-fri" 
      duration: 24h 
      reasons: ["underutilized"] 
    - nodes: "100%" 
      schedule: "* * * * sat"
      duration: 48h
      reasons: ["underutilized"]
```

#### Which Equation Best Covers all the scenarios? 
### Q: How should we handle an unspecified reason when others are specified? 
```yaml
budgets: 
  - nodes: 10
    reasons: [Drifted, Underutilized]
    schedule: "* * * * *"
  - nodes: 5 
    reasons: [Empty] 
    schedule: "* * * * *" 
```
In the case of a budget like above, default is undefined. Should karpenter assume the user doesn't want to disrupt any other reasons? Or should we assume that if a default is unspecified, they want us to disrupt anyway?  

> The intuitive options if there is no active default budget is to allow disruption of either 0 or total number of nodes(meaning unbounded disruption).

Reviewer 5: I'd expect us to use the current karpenter default, in a case like this that doesn't have a default specified. 

Karpenter's current handling of this is `10%`:
https://karpenter.sh/v0.35/concepts/disruption/#disruption-budgets
![image](https://github.com/kubernetes-sigs/karpenter/assets/33269602/021e54cf-d159-4799-8087-d7fa72d8ec1d)

I think unbound default is both unexpected, and could result in concerning scenarios for unaware users. `10-30`% seems most reasonable, so sticking with the current 10% default seems good to me.
